README.md
Mobile secure keyboard to prevent KeyLogger attack and screen capture.

## Screenshots
| Alphanumeric | Numeric |
|---|---|
| <img src="https://user-images.githubusercontent.com/47127353/111059243-52fbc980-84d7-11eb-9260-01b6b7b5909c.png" width="200"> | <img src="https://user-images.githubusercontent.com/47127353/111059256-7888d300-84d7-11eb-8797-89a69337d8ad.png" width="200"> |

## Getting started

To use this plugin, add `flutter_secure_keyboard` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/). For example:

```yaml
dependencies:
  flutter_secure_keyboard: ^2.1.2
```

## Examples

```dart
class WithSecureKeyboardExample extends StatefulWidget {
  @override
  _WithSecureKeyboardExampleState createState() => _WithSecureKeyboardExampleState();
}

class _WithSecureKeyboardExampleState extends State<WithSecureKeyboardExample> {
  final secureKeyboardController = SecureKeyboardController();

  final passwordEditor = TextEditingController();
  final passwordTextFieldFocusNode = FocusNode();

  final pinCodeEditor = TextEditingController();
  final pinCodeTextFieldFocusNode = FocusNode();

  @override
  Widget build(BuildContext context) {
    // We recommend that you set the secure keyboard to the top 
    // of the build function so that it can be seen properly.
    return WithSecureKeyboard(
      controller: secureKeyboardController,
      child: Scaffold(
        appBar: AppBar(title: Text('with_secure_keyboard_example')),
        body: _buildContentView()
      ),
    );
  }

  Widget _buildContentView() {
    // We recommend using the ListView widget to prevent widget overflow.
    return ListView(
      padding: const EdgeInsets.all(8.0),
      children: [
        Padding(
          padding: const EdgeInsets.only(bottom: 12.0),
          child: _buildPasswordTextField()
        ),
        _buildPinCodeTextField()
      ],
    );
  }

  Widget _buildPasswordTextField() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('Password'),
        TextFormField(
          controller: passwordEditor,
          focusNode: passwordTextFieldFocusNode,
          // We recommended to set false to prevent the soft keyboard from opening.
          enableInteractiveSelection: false,
          obscureText: true,
          onTap: () {
            secureKeyboardController.show(
              type: SecureKeyboardType.ALPHA_NUMERIC,
              textFieldFocusNode: passwordTextFieldFocusNode,
              initText: passwordEditor.text,
              hintText: 'password',
              // Use onCharCodesChanged to have text entered in real time.
              onCharCodesChanged: (List<int> charCodes) {
                passwordEditor.text = String.fromCharCodes(charCodes);
              }
            );
          },
        ),
      ],
    );
  }

  Widget _buildPinCodeTextField() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('PinCode'),
        TextFormField(
          controller: pinCodeEditor,
          focusNode: pinCodeTextFieldFocusNode,
          // We recommended to set false to prevent the soft keyboard from opening.
          enableInteractiveSelection: false,
          obscureText: true,
          onTap: () {
            secureKeyboardController.show(
              type: SecureKeyboardType.NUMERIC,
              textFieldFocusNode: pinCodeTextFieldFocusNode,
              initText: pinCodeEditor.text,
              hintText: 'pinCode',
              // Use onDoneKeyPressed to allow text to be entered when you press the done key,
              // or to do something like encryption.
              onDoneKeyPressed: (List<int> charCodes) {
                pinCodeEditor.text = String.fromCharCodes(charCodes);
              }
            );
          },
        ),
      ],
    );
  }
}
```

## Package

* [WithSecureKeyboard] - A widget that implements a secure keyboard with controller.
* [SecureKeyboardController] - Controller to check or control the state of the secure keyboard.

**Note:** The parameters marked with an asterisk(*) are required.

### WithSecureKeyboard

| Parameter | Description |
|---|---|
| `controller`* | Controller for controlling the secure keyboard. |
| `child`* | A widget to have a secure keyboard. |
| `keyboardHeight` | Parameter to set the keyboard height. <br> Default value is `280.0`. |
| `keyRadius` | Set the radius of the keyboard key. <br> Default value is `4.0`. |
| `keySpacing` | Set the spacing between keyboard keys. <br> Default value is `1.4`. |
| `keyInputMonitorPadding` | Set the padding of the key input monitor. <br> Default value is `EdgeInsets.only(left: 10.0, right: 5.0)`. |
| `keyboardPadding` | Set the padding of the keyboard. <br> Default value is `EdgeInsets.symmetric(horizontal: 5.0)`. |
| `backgroundColor` | Parameter to set the keyboard background color. <br> Default value is `Color(0xFF0A0A0A)`. |
| `stringKeyColor` | Parameter to set keyboard string key(alphanumeric, numeric..) color. <br> Default value is `Color(0xFF313131)`. |
| `actionKeyColor` | Parameter to set keyboard action key(shift, backspace, clear..) color. <br> Default value is `Color(0xFF222222)`. |
| `doneKeyColor` | Parameter to set keyboard done key color. <br> Default value is `Color(0xFF1C7CDC)`. |
| `activatedKeyColor` | Set the color to display when activated with the shift action key. If the value is null, `doneKeyColor` is used. |
| `keyTextStyle` | Parameter to set keyboard key text style. <br> Default value is `TextStyle(color: Colors.white, fontSize: 16.0, fontWeight: FontWeight.bold)`. |
| `inputTextStyle` | Parameter to set keyboard input text style. <br> Default value is `TextStyle(color: Colors.white, fontSize: 18.0, fontWeight: FontWeight.bold)`. |
| `screenCaptureDetectedAlertTitle` | Security Alert title, only works on ios. |
| `screenCaptureDetectedAlertMessage` | Security Alert message, only works on ios |
| `screenCaptureDetectedAlertActionTitle` | Security Alert actionTitle, only works on ios. |

### SecureKeyboardController

| Function | Description |
|---|---|
| `isShowing` | Whether the secure keyboard is open. |
| `show` | Show a secure keyboard. |
| `hide` | Hide a secure keyboard. |

### SecureKeyboardController.show()

| Parameter | Description |
|---|---|
| `type`* | Specifies the secure keyboard type. |
| `textFieldFocusNode` | The `FocusNode` that will receive focus on. |
| `initText` | Set the initial value of the input text. |
| `hintText` | The hint text to display when the input text is empty. |
| `inputTextLengthSymbol` | Set the symbol to use when displaying the input text length. |
| `doneKeyText` | Set the done key text. |
| `clearKeyText` | Set the clear key text. |
| `obscuringCharacter` | Set the secure character to hide the input text. <br> Default value is `•`. |
| `maxLength` | Set the maximum length of text that can be entered. |
| `alwaysCaps` | Whether to always display uppercase characters. <br> Default value is `false`. |
| `obscureText` | Whether to hide input text as secure characters. <br> Default value is `true`. |
| `shuffleNumericKey` | Whether to shuffle the position of the numeric keys. <br> Default value is `true`. |
| `hideKeyInputMonitor` | Whether to hide the key input monitor. <br> Default value is `false`. |
| `disableKeyBubble` | Whether to disable the key bubble. <br> Default value is `false`. |
| `onKeyPressed` | Called when the key is pressed. |
| `onCharCodesChanged` | Called when the character codes changed. |
| `onDoneKeyPressed` | Called when the done key is pressed. |
| `onCloseKeyPressed` | Called when the close key is pressed. |


CHANGELOG.md
## 2.1.2

* Fixed an issue that lost focus when switching soft keyboard.

## 2.1.1

* Add `hideKeyInputMonitor` option.
* Add `disableKeyBubble` option.

## 2.1.0

* Improve performance.
* Add `Key Bubble` to check the string key pressed.

## 2.0.0

* Migrate null safety.
* Upgrade `flutter_keyboard_visibility` plugin.
* Remove `type` function in `SecureKeyboardController`.

## 1.1.0

* Upgrade `flutter_keyboard_visibility` plugin.
* Change the value of `SecureKeyboardKeyAction` to uppercase.
* Change the value of `SecureKeyboardKeyType` to uppercase.
* Change the value of `SecureKeyboardType` to uppercase.
* Add `shuffleNumericKey` option to ask for shuffle of numeric keys.
* Add `keyRadius` option to set the radius of the key button.
* Add `keySpacing` option to set the spacing between key buttons.
* Add `keyInputMonitorPadding`, `keyboardPadding` option to cope with edge screen.

## 1.0.6

* Example updates.
* Documentation updates.
* Add code so that the page does not close when the back button on the soft navigation bar is pressed.

## 1.0.5

* README updates.
* Example updates.
* Remove `onDoneKeyPressed` required annotation.

## 1.0.4

* Add `activatedKeyColor` param.

## 1.0.3

* Change action key name from confirm to done.
* Fixed an error that occur when setting the key text.

## 1.0.2

* Add `screenCaptureDetectedAlertTitle` param.
* Add `screenCaptureDetectedAlertMessage` param.
* Add `screenCaptureDetectedAlertActionTitle` param.

## 1.0.1+3

* Update documentation.

## 1.0.1+2

* README updates.
* Change package name and structure.

## 1.0.1

* Fixed an issue that caused the input text to overflow.
* Added option to set maximum length of input text.

## 1.0.0+2

* Resize the close and view buttons.
* Fixed an issue where the button function did not work when dragging the view button.

## 1.0.0

* Initial release.


lib/src/secure_keyboard.dart
import 'dart:async';
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_action.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_generator.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_type.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_type.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key.dart';

/// Height of widget showing input text.
const double kKeyInputMonitorHeight = 50.0;

/// Keyboard default height.
const double kKeyboardDefaultHeight = 280.0;

/// Keyboard key default radius.
const double kKeyboardKeyDefaultRadius = 4.0;

/// Keyboard key default spacing.
const double kKeyboardKeyDefaultSpacing = 1.4;

/// Speed ​​of erasing input text when holding backspace.
const int kBackspaceEventDelay = 100;

/// Keyboard default background color.
const Color kKeyboardDefaultBackgroundColor = Color(0xFF0A0A0A);

/// Keyboard default string key color.
const Color kKeyboardDefaultStringKeyColor = Color(0xFF313131);

/// Keyboard default action key color.
const Color kKeyboardDefaultActionKeyColor = Color(0xFF222222);

/// Keyboard default done key color.
const Color kKeyboardDefaultDoneKeyColor = Color(0xFF1C7CDC);

/// Keyboard default key text style.
const TextStyle kKeyboardDefaultKeyTextStyle = TextStyle(
    color: Colors.white, fontSize: 16.0, fontWeight: FontWeight.bold);

/// Keyboard default input text style.
const TextStyle kKeyboardDefaultInputTextStyle = TextStyle(
    color: Colors.white, fontSize: 18.0, fontWeight: FontWeight.bold);

/// Key input monitor default padding.
const EdgeInsetsGeometry kKeyInputMonitorDefaultPadding =
EdgeInsets.only(left: 10.0, right: 5.0);

/// Keyboard default padding.
const EdgeInsetsGeometry kKeyboardDefaultPadding =
EdgeInsets.symmetric(horizontal: 5.0);

/// String key touch start callback function.
typedef OnStringKeyTouchStart = void Function(
    String keyText, Offset position, BoxConstraints constraints);

/// String key touch end callback function.
typedef OnStringKeyTouchEnd = void Function();

/// Widget that implements a secure keyboard.
class SecureKeyboard extends StatefulWidget {
  /// Specifies the secure keyboard type.
  final SecureKeyboardType type;

  /// Called when the key is pressed.
  final ValueChanged<SecureKeyboardKey> onKeyPressed;

  /// Called when the character codes changed.
  final ValueChanged<List<int>> onCharCodesChanged;

  /// Called when the done key is pressed.
  final ValueChanged<List<int>> onDoneKeyPressed;

  /// Called when the close key is pressed.
  final VoidCallback onCloseKeyPressed;

  /// Called when the string key touch start.
  final OnStringKeyTouchStart onStringKeyTouchStart;

  /// Called when the string key touch end.
  final OnStringKeyTouchEnd onStringKeyTouchEnd;

  /// Set the initial value of the input text.
  final String? initText;

  /// The hint text to display when the input text is empty.
  final String? hintText;

  /// Set the symbol to use when displaying the input text length.
  final String? inputTextLengthSymbol;

  /// Set the done key text.
  final String? doneKeyText;

  /// Set the clear key text.
  final String? clearKeyText;

  /// Set the secure character to hide the input text.
  /// Default value is `•`.
  final String obscuringCharacter;

  /// Set the maximum length of text that can be entered.
  final int? maxLength;

  /// Whether to always display uppercase characters.
  /// Default value is `false`.
  final bool alwaysCaps;

  /// Whether to hide input text as secure characters.
  /// Default value is `true`.
  final bool obscureText;

  /// Whether to shuffle the position of the numeric keys.
  /// Default value is `true`.
  final bool shuffleNumericKey;

  /// Whether to hide the key input monitor.
  /// Default value is `false`.
  final bool hideKeyInputMonitor;

  /// Parameter to set the keyboard height.
  /// Default value is `280.0`.
  final double height;

  /// Set the radius of the keyboard key.
  /// Default value is `4.0`.
  final double keyRadius;

  /// Set the spacing between keyboard keys.
  /// Default value is `1.4`.
  final double keySpacing;

  /// Set the padding of the key input monitor.
  /// Default value is `EdgeInsets.only(left: 10.0, right: 5.0)`.
  final EdgeInsetsGeometry keyInputMonitorPadding;

  /// Set the padding of the keyboard.
  /// Default value is `EdgeInsets.symmetric(horizontal: 5.0)`.
  final EdgeInsetsGeometry keyboardPadding;

  /// Parameter to set the keyboard background color.
  /// Default value is `Color(0xFF0A0A0A)`.
  final Color backgroundColor;

  /// Parameter to set keyboard string key(alphanumeric, numeric..) color.
  /// Default value is `Color(0xFF313131)`.
  final Color stringKeyColor;

  /// Parameter to set keyboard action key(shift, backspace, clear..) color.
  /// Default value is `Color(0xFF222222)`.
  final Color actionKeyColor;

  /// Parameter to set keyboard done key color.
  /// Default value is `Color(0xFF1C7CDC)`.
  final Color doneKeyColor;

  /// Set the color to display when activated with the shift action key.
  /// If the value is null, `doneKeyColor` is used.
  final Color? activatedKeyColor;

  /// Parameter to set keyboard key text style.
  /// Default value is `TextStyle(color: Colors.white, fontSize: 16.0, fontWeight: FontWeight.bold)`.
  final TextStyle keyTextStyle;

  /// Parameter to set keyboard input text style.
  /// Default value is `TextStyle(color: Colors.white, fontSize: 18.0, fontWeight: FontWeight.bold)`.
  final TextStyle inputTextStyle;

  /// Security Alert title, only works on ios.
  final String? screenCaptureDetectedAlertTitle;

  /// Security Alert message, only works on ios.
  final String? screenCaptureDetectedAlertMessage;

  /// Security Alert actionTitle, only works on ios.
  final String? screenCaptureDetectedAlertActionTitle;

  SecureKeyboard({
    Key? key,
    required this.type,
    required this.onKeyPressed,
    required this.onCharCodesChanged,
    required this.onDoneKeyPressed,
    required this.onCloseKeyPressed,
    required this.onStringKeyTouchStart,
    required this.onStringKeyTouchEnd,
    this.initText,
    this.hintText,
    this.inputTextLengthSymbol,
    this.doneKeyText,
    this.clearKeyText,
    this.obscuringCharacter = '•',
    this.maxLength,
    this.alwaysCaps = false,
    this.obscureText = true,
    this.shuffleNumericKey = true,
    this.hideKeyInputMonitor = false,
    this.height = kKeyboardDefaultHeight,
    this.keyRadius = kKeyboardKeyDefaultRadius,
    this.keySpacing = kKeyboardKeyDefaultSpacing,
    this.keyInputMonitorPadding = kKeyInputMonitorDefaultPadding,
    this.keyboardPadding = kKeyboardDefaultPadding,
    this.backgroundColor = kKeyboardDefaultBackgroundColor,
    this.stringKeyColor = kKeyboardDefaultStringKeyColor,
    this.actionKeyColor = kKeyboardDefaultActionKeyColor,
    this.doneKeyColor = kKeyboardDefaultDoneKeyColor,
    this.activatedKeyColor,
    this.keyTextStyle = kKeyboardDefaultKeyTextStyle,
    this.inputTextStyle = kKeyboardDefaultInputTextStyle,
    this.screenCaptureDetectedAlertTitle,
    this.screenCaptureDetectedAlertMessage,
    this.screenCaptureDetectedAlertActionTitle
  })  : assert(obscuringCharacter.isNotEmpty),
        super(key: key);

  @override
  _SecureKeyboardState createState() => _SecureKeyboardState();
}

class _SecureKeyboardState extends State<SecureKeyboard> {
  final _methodChannel = const MethodChannel('flutter_secure_keyboard');

  final _definedKeyRows = <List<SecureKeyboardKey>>[];
  final _specialKeyRows = <List<SecureKeyboardKey>>[];

  final _charCodesController = StreamController<List<int>>.broadcast();
  final _charCodes = <int>[];
  
  Timer? _backspaceEventGenerator;

  bool _isViewEnabled = false;
  bool _isShiftEnabled = false;
  bool _isSpecialCharsEnabled = false;
  
  void _initVariables() {
    _isViewEnabled = false;
    _isShiftEnabled = false;
    _isSpecialCharsEnabled = false;

    _definedKeyRows.clear();
    _specialKeyRows.clear();
    _charCodes.clear();
    if (widget.initText != null)
      _charCodes.addAll(widget.initText!.codeUnits);

    final keyGenerator = SecureKeyboardKeyGenerator.instance;
    if (widget.type == SecureKeyboardType.NUMERIC)
      _definedKeyRows.addAll(keyGenerator.getNumericKeyRows(widget.shuffleNumericKey));
    else
      _definedKeyRows.addAll(keyGenerator.getAlphanumericKeyRows(widget.shuffleNumericKey));
    _specialKeyRows.addAll(SecureKeyboardKeyGenerator.instance.getSpecialCharsKeyRows());
  }

  void _notifyCharCodesChanged() {
    _charCodesController.sink.add(_charCodes);
  }

  void _onKeyPressed(SecureKeyboardKey key) {
    if (key.type == SecureKeyboardKeyType.STRING) {
      // The length of `charCodes` cannot exceed `maxLength`.
      final currMaxLength = widget.maxLength;
      if (currMaxLength != null && (currMaxLength <= _charCodes.length))
        return;

      final keyText = (_isShiftEnabled || widget.alwaysCaps)
          ? key.capsText
          : key.text;
      if (keyText == null) return;

      _charCodes.add(keyText.codeUnits.first);
      _notifyCharCodesChanged();
      widget.onCharCodesChanged(_charCodes);
    } else if (key.type == SecureKeyboardKeyType.ACTION) {
      switch (key.action) {
        // Backspace
        case SecureKeyboardKeyAction.BACKSPACE:
          if (_charCodes.isNotEmpty) {
            _charCodes.removeLast();
            _notifyCharCodesChanged();
            widget.onCharCodesChanged(_charCodes);
          }
          break;
          
        // Done
        case SecureKeyboardKeyAction.DONE:
          widget.onDoneKeyPressed(_charCodes);
          break;
          
        // Clear
        case SecureKeyboardKeyAction.CLEAR:
          _charCodes.clear();
          _notifyCharCodesChanged();
          widget.onCharCodesChanged(_charCodes);
          break;
          
        // Shift
        case SecureKeyboardKeyAction.SHIFT:
          if (!widget.alwaysCaps)
            setState(() {
              _isShiftEnabled = !_isShiftEnabled;
            });
          break;
          
        // SpecialChars
        case SecureKeyboardKeyAction.SPECIAL_CHARACTERS:
          setState(() {
            _isSpecialCharsEnabled = !_isSpecialCharsEnabled;
          });
          break;

        default:
          return;
      }
    }

    widget.onKeyPressed(key);
  }

  @override
  void didUpdateWidget(covariant SecureKeyboard oldWidget) {
    super.didUpdateWidget(oldWidget);
    _initVariables();
  }

  @override
  void initState() {
    super.initState();
    _methodChannel.invokeMethod('secureModeOn', {
      'screenCaptureDetectedAlertTitle': widget.screenCaptureDetectedAlertTitle,
      'screenCaptureDetectedAlertMessage': widget.screenCaptureDetectedAlertMessage,
      'screenCaptureDetectedAlertActionTitle': widget.screenCaptureDetectedAlertActionTitle
    });
    _initVariables();
  }

  @override
  Widget build(BuildContext context) {
    double height;
    Widget keyInputMonitor;
    if (widget.hideKeyInputMonitor) {
      height = widget.height + 5.0;
      keyInputMonitor = Container();
    } else {
      height = widget.height + kKeyInputMonitorHeight;
      keyInputMonitor = Padding(
        padding: widget.keyInputMonitorPadding,
        child: _buildKeyInputMonitor()
      );
    }

    final keyRows = _isSpecialCharsEnabled
        ? _specialKeyRows
        : _definedKeyRows;
    final keyboard = Padding(
      padding: widget.keyboardPadding,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.end,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: _buildKeyboardKey(keyRows)
      ),
    );

    return WillPopScope(
      onWillPop: () async {
        widget.onCloseKeyPressed();
        return false;
      },
      child: Container(
        width: MediaQuery.of(context).size.width,
        height: height,
        color: widget.backgroundColor,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.end,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            keyInputMonitor,
            keyboard
          ]
        ),
      ),
    );
  }

  @override
  void dispose() {
    _methodChannel.invokeMethod('secureModeOff');
    _charCodesController.close();
    super.dispose();
  }

  Widget _buildKeyInputMonitor() {
    return StreamBuilder<List<int>>(
      stream: _charCodesController.stream.asBroadcastStream(
        onCancel: (subscription) => subscription.cancel()
      ),
      initialData: _charCodes,
      builder: (context, snapshot) =>
          _buildKeyInputMonitorLayout(snapshot.data ?? _charCodes)
    );
  }

  Widget _buildKeyInputMonitorLayout(List<int> charCodes) {
    String secureText;
    TextStyle secureTextStyle;

    if (charCodes.isNotEmpty) {
      if (widget.obscureText && !_isViewEnabled) {
        secureText = '';
        for (int i=0; i<charCodes.length; i++) {
          if (i == charCodes.length - 1)
            secureText += String.fromCharCode(charCodes[i]);
          else
            secureText += widget.obscuringCharacter;
        }
      } else {
        secureText = String.fromCharCodes(charCodes);
      }

      secureTextStyle = widget.inputTextStyle;
    } else {
      secureText = widget.hintText ?? '';
      secureTextStyle = widget.inputTextStyle.copyWith(
          color: widget.inputTextStyle.color?.withOpacity(0.5));
    }

    String? lengthSymbol = widget.inputTextLengthSymbol;
    if (lengthSymbol == null)
      lengthSymbol = (Platform.localeName == 'ko_KR') ? '자' : 'digit';
    final lengthText = '${charCodes.length}$lengthSymbol';

    return SizedBox(
      height: kKeyInputMonitorHeight,
      child: Row(
        children: [
          Expanded(
            child: Text(
              secureText,
              style: secureTextStyle,
              maxLines: 1,
              overflow: TextOverflow.ellipsis
            )
          ),
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: 5.0),
            child: Text(lengthText, style: widget.keyTextStyle)
          ),
          (widget.obscureText)
              ? _buildViewButton()
              : SizedBox(),
          _buildCloseButton()
        ],
      ),
    );
  }

  Widget _buildViewButton() {
    return Container(
      width: kKeyInputMonitorHeight / 1.4,
      height: kKeyInputMonitorHeight / 1.4,
      margin: const EdgeInsets.only(left: 1.5),
      child: Material(
        color: Colors.transparent,
        child: InkWell(
          borderRadius: BorderRadius.circular(24.0),
          onTap: () {

          },
          child: GestureDetector(
            onPanDown: (_) => setState(() => _isViewEnabled = true),
            onPanEnd: (_) => setState(() => _isViewEnabled = false),
            onPanCancel: () => setState(() => _isViewEnabled = false),
            child: Icon(Icons.remove_red_eye, color: widget.keyTextStyle.color)
          ),
        ),
      ),
    );
  }

  Widget _buildCloseButton() {
    return Container(
      width: kKeyInputMonitorHeight / 1.4,
      height: kKeyInputMonitorHeight / 1.4,
      margin: const EdgeInsets.only(right: 1.5),
      child: Material(
        color: Colors.transparent,
        child: InkWell(
          borderRadius: BorderRadius.circular(24.0),
          onTap: widget.onCloseKeyPressed,
          child: Icon(Icons.close, color: widget.keyTextStyle.color)
        ),
      ),
    );
  }

  List<Widget> _buildKeyboardKey(List<List<SecureKeyboardKey>> keyRows) {
    return List.generate(keyRows.length, (int rowNum) {
      return Material(
        color: Colors.transparent,
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceAround,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: List.generate(keyRows[rowNum].length, (int keyNum) {
            final key = keyRows[rowNum][keyNum];
            
            switch (key.type) {
              case SecureKeyboardKeyType.STRING:
                return _buildStringKey(key, keyRows.length);
              case SecureKeyboardKeyType.ACTION:
                return _buildActionKey(key, keyRows.length);
              default:
                throw Exception('Unknown key type.');
            }
          })
        ),
      );
    });
  }
  
  Widget _buildStringKey(SecureKeyboardKey key, int keyRowsLength) {
    final keyText = (_isShiftEnabled || widget.alwaysCaps)
        ? key.capsText ?? ''
        : key.text ?? '';
    final stringKeyChild = Text(keyText, style: widget.keyTextStyle);
    final globalKey = GlobalKey();

    return Expanded(
      child: Container(
        key: globalKey,
        height: widget.height / keyRowsLength,
        padding: EdgeInsets.all(widget.keySpacing),
        child: LayoutBuilder(
          builder: (context, constraints) {
            final stringKey = Material(
              borderRadius: BorderRadius.circular(widget.keyRadius),
              color: widget.stringKeyColor,
              child: InkWell(
                borderRadius: BorderRadius.circular(widget.keyRadius),
                onTap: () => _onKeyPressed(key),
                child: Center(child: stringKeyChild)
              ),
            );

            if (widget.type == SecureKeyboardType.ALPHA_NUMERIC)
              return GestureDetector(
                onPanDown: (_) {
                  final renderObject = globalKey.currentContext?.findRenderObject();
                  if ((renderObject != null) && (renderObject is RenderBox)) {
                    final position = renderObject.localToGlobal(Offset.zero);
                    widget.onStringKeyTouchStart(keyText, position, constraints);
                  }
                },
                onPanEnd: (_) => widget.onStringKeyTouchEnd(),
                onPanCancel: () => widget.onStringKeyTouchEnd(),
                child: stringKey
              );

            return stringKey;
          }
        ),
      ),
    );
  }

  Widget _buildActionKey(SecureKeyboardKey key, int keyRowsLength) {
    Widget actionKeyChild;
    switch (key.action ?? SecureKeyboardKeyAction.BLANK) {
      case SecureKeyboardKeyAction.BACKSPACE:
        actionKeyChild = GestureDetector(
          onLongPress: () {
            final delay = Duration(milliseconds: kBackspaceEventDelay);
            _backspaceEventGenerator = Timer.periodic(delay, (_) => _onKeyPressed(key));
          },
          onLongPressUp: () {
            _backspaceEventGenerator?.cancel();
            _backspaceEventGenerator = null;
          },
          child: Icon(Icons.backspace, color: widget.keyTextStyle.color)
        );
        break;
      case SecureKeyboardKeyAction.SHIFT:
        actionKeyChild = Icon(Icons.arrow_upward, color: widget.keyTextStyle.color);
        break;
      case SecureKeyboardKeyAction.CLEAR:
        String? keyText = widget.clearKeyText;
        if (keyText == null || keyText.isEmpty)
          keyText = (Platform.localeName == 'ko_KR') ? '초기화' : 'Clear';

        actionKeyChild = Text(keyText, style: widget.keyTextStyle);
        break;
      case SecureKeyboardKeyAction.DONE:
        String? keyText = widget.doneKeyText;
        if (keyText == null || keyText.isEmpty)
          keyText = (Platform.localeName == 'ko_KR') ? '입력완료' : 'Done';

        actionKeyChild = Text(keyText, style: widget.keyTextStyle);
        break;
      case SecureKeyboardKeyAction.SPECIAL_CHARACTERS:
        actionKeyChild = Text(
          _isSpecialCharsEnabled
              ? (_isShiftEnabled ? 'ABC' : 'abc')
              : '!@#',
          style: widget.keyTextStyle
        );
        break;
      case SecureKeyboardKeyAction.BLANK:
        return Expanded(child: SizedBox());
    }

    Color keyColor;
    if (key.action == SecureKeyboardKeyAction.DONE)
      keyColor = widget.doneKeyColor;
    else if (key.action == SecureKeyboardKeyAction.SHIFT && _isShiftEnabled)
      keyColor = widget.activatedKeyColor ?? widget.doneKeyColor;
    else
      keyColor = widget.actionKeyColor;

    return Expanded(
      child: Container(
        height: widget.height / keyRowsLength,
        padding: EdgeInsets.all(widget.keySpacing),
        child: Material(
          borderRadius: BorderRadius.circular(widget.keyRadius),
          color: keyColor,
          child: InkWell(
            borderRadius: BorderRadius.circular(widget.keyRadius),
            onTap: () => _onKeyPressed(key),
            child: Center(child: actionKeyChild)
          ),
        ),
      ),
    );
  }
}
